File: Android/app/src/main/java/com/example/lpstudio/UpdateCoordinator.java
package com.example.lpstudio;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;
import com.example.lpstudio.AndroidUpdateService;

public class UpdateCoordinator {
    private static final String TAG = "UpdateCoordinator";

    private final Context context;
    private final Handler mainHandler;
    private final AndroidUpdateService updateService;
    private UpdateListener listener;

    public UpdateCoordinator(Context context) {
        this(context, "Albel-coder", "LegoPrinter");
    }

    public UpdateCoordinator(Context context, String owner, String repo) {
        // Используем applicationContext для предотвращения утечек памяти
        this.context = context.getApplicationContext();
        this.mainHandler = new Handler(Looper.getMainLooper());
        this.updateService = new AndroidUpdateService(this.context, owner, repo);

        updateService.setUpdateCallback(new AndroidUpdateService.UpdateCallback() {
            @Override
            public void onError(String message) {
                mainHandler.post(() -> {
                    Toast.makeText(context, message, Toast.LENGTH_LONG).show();
                    if (listener != null) {
                        listener.onError(message);
                    }
                });
            }

            @Override
            public void onInfo(String message) {
                mainHandler.post(() -> {
                   Toast.makeText(context, message, Toast.LENGTH_LONG).show();
                });
            }
        });
    }

    public void setUpdateListener(UpdateListener listener) {
        this.listener = listener;
    }

    // Убираем лишний Thread - делегируем всю работу AndroidUpdateService
    public void checkForUpdates(boolean force) {
        updateService.checkForUpdates(new AndroidUpdateService.UpdateCheckCallback() {
            @Override
            public void onResult(AndroidUpdateService.UpdateInfo updateInfo) {
                mainHandler.post(() -> {
                    if (listener != null && updateInfo.isAvailable) {
                        listener.onUpdateAvailable(updateInfo);
                    }
                });
                Log.d("UpdateCheck", "Current: " + updateInfo.currentVersion + "/" + updateInfo.currentVersionCode +
                        ", Latest: " + updateInfo.latestVersion + "/" + updateInfo.latestBuild);
            }

            @Override
            public void onError(String error) {
                Log.e(TAG, "Update check error: " + error);
                mainHandler.post(() -> {
                    if (listener != null) {
                        listener.onError(error);
                    }
                });
            }
        });

    }

    public void downloadUpdate(AndroidUpdateService.UpdateInfo info) {
        // Оставляем Thread только для загрузки
        new Thread(() -> {
            try {
                long downloadId = updateService.downloadUpdate(info);
                Log.i(TAG, "Download started with ID: " + downloadId);

                mainHandler.post(() -> {
                    if (listener != null) {
                        listener.onDownloadStarted(info);
                    }
                });

            } catch (Exception e) {
                Log.e(TAG, "Download error: " + e.getMessage());
                mainHandler.post(() -> {
                    if (listener != null) {
                        listener.onError("Ошибка загрузки: " + e.getMessage());
                    }
                });
            }
        }).start();
    }

    public void skipVersion(int versionCode) {
        updateService.skipVersion(versionCode);
    }

    public void postponeUpdate(int hours) {
        updateService.postponeUpdate(hours);
    }

    public void cleanup() {
        updateService.cleanup();
    }

    public interface UpdateListener {
        void onUpdateAvailable(AndroidUpdateService.UpdateInfo info);
        void onDownloadStarted(AndroidUpdateService.UpdateInfo info);
        void onError(String message);
    }


}

File: Android/app/src/main/java/com/example/lpstudio/UpdateCoordinator.kt
package com.example.lpstudio

import android.content.Context
import com.example.lpstudio.AndroidUpdateService
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class UpdateCoordinator(private val context: Context) {

    private val updateService = AndroidUpdateService(context)
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    private val _updateState = MutableStateFlow<UpdateState>(UpdateState.Idle)
    val updateState: StateFlow<UpdateState> =_updateState

    sealed class UpdateState {
        object Idle : UpdateState()
        object Checking : UpdateState()
        data class Available(val info: AndroidUpdateService.UpdateInfo) : UpdateState()
        data class Downloading(val progress: Float) : UpdateState()
        data class Installing(val info: AndroidUpdateService.UpdateInfo) : UpdateState()
        data class Error(val message: String) : UpdateState()
    }

    fun checkForUpdates(force: Boolean = false) {
        scope.launch {
            _updateState.value = UpdateState.Checking

            try {
                val updateInfo = withContext(Dispatchers.IO) {
                    updateService.checkForUpdates(force)
                }

                if (updateInfo.isAvailable) {
                    _updateState.value = UpdateState.Available(updateInfo)
                }
                else {
                    _updateState.value = UpdateState.Idle
                }
            }
            catch(e: Exception) {
                _updateState.value = UpdateState.Error("Error checking: ${e.message}")
            }
        }
    }

    fun downloadUpdate(info: AndroidUpdateService.UpdateInfo) {
        scope.launch {
            try {
                val downloadId = withContext(Dispatchers.IO) {
                    updateService.downloadAndInstall(info)
                }
                _updateState.value = UpdateState.Downloading(0f)
            }
            catch(e: Exception) {
                _updateState.value = UpdateState.Error("Error downloading: ${e.message}")
            }
        }
    }

    fun skipVersion(versionCode: Int) {
        updateService.skipVersion(versionCode)
        _updateState.value = UpdateState.Idle
    }

    fun postponeUpdate(hours: Int) {
        updateService.postponeUpdate(hours)
        _updateState.value = UpdateState.Idle
    }

    fun cleanup() {
        scope.cancel()
        updateService.cleanup()
    }
}